packages/navigator/src/browser/navigator-widget.tsx

```diff
@@ -187,4 +187,29 @@ export class FileNavigatorWidget extends FileTreeWidget {
             </div>
         </div>;
     }
+
+    protected getDecorationData<K extends keyof TreeDecoration.Data>(node: TreeNode, key: K): TreeDecoration.Data[K][] {
+        const decorations = super.getDecorationData(node, key);
+        if (WorkspaceRootNode.is(node)) {
+            const uri = UriSelection.getUri(node);
+            const roots = this.workspaceService.tryGetRoots();
+            if (uri && roots.findIndex(r => r.uri === uri.toString()) < 0) {
+                if (key === 'iconOverlay') {
+                    decorations.push({
+                        position: TreeDecoration.IconOverlayPosition.BOTTOM_RIGHT,
+                        icon: 'exclamation-circle',
+                        color: 'var(--theia-warn-color0)',
+                        background: {
+                            shape: 'circle',
+                            color: 'var(--theia-layout-color0)'
+                        }
+                    });
+                }
+                if (key === 'tooltip') {
+                    decorations.push('Workspace root cannot be found');
```

Comment:
I wonder should not we use problem markers API for it?

Reply:
it was just a question :) I was wondering whether we want to see such errors in the problem view as well. Let's leave as it is.

Reply:
It would be indeed better to have it using the Marker api. For instance, we could later have quick-fixes associated with them. But I see it is a bit complicated, so maybe we go back to this once there is a good reason.

Reply:
@akosyakov 
I updated the WorkspaceService API. Now it exposes all roots listed in the workspace file (regardless of existence). 

This change introduces a technical difficulty for me:
in the `NavigatorWidget`, I was counting on `WorkspaceService.tryGetAllRoots()` & `WorkspaceService.tryGetRoots()` to tell which roots are invalid. Now `WorkspaceService.getRoots()` returns all roots, including the invalid ones. To achieve the same goal, I would need `async filesystem.exist()` to validate, which makes a big part of the tree rendering logic async as well. 

a few things I could do:
1) use a separate decorator. this is what i used to do 
2) use the makers API. This requires me to import makers in either `workspace-service` or `navigator`, and results in circular dependency. I am not sure if there is a way to easily work around it. 
3) change the problem markers decorator directly. but this is ugly as we don't want to use the same thing for two different purposes.
4) do not add any icons or tooltips to the invalid roots. User would find no nodes under the invalid root, and figure out why. 

None of the above 4 is ideal ... do you have any suggestions ?

Reply:
@elaihau 
> [...] should I put the code in the navigator extension? It looks difficult, as markers is already dependent on the navigator

Would it prevent the circular dependency issue if the workspace service took care of it? i.e. check the existence of the various roots, and create/manage problem markers about missing ones. 

Are there other impediments, about using markers?

Reply:
@elaihau it is indeed complicated, we can do it later